feat!: footer legal links

asadali145 opened the pull request:
## Related Issues
Product Review Issue: https://github.com/openedx/platform-roadmap/issues/324
https://github.com/mitodl/mitxonline/issues/223

## Description
Legacy LMS provides a way to configure the footer links but the MFE footer has no support for legal links. This PR adds support to configure custom legal links in the MFE footer.

## Testing

- This PR can be tested with the Accounts MFE https://github.com/openedx/frontend-app-account.
- Test without adding the links. The footer should be the same as before.
- Add config links in .env.development:
  - ACCESSIBILITY_URL
  - ABOUT_US_URL
  - HONOR_CODE_URL
  - CONTACT_URL
  - SUPPORT_CENTER_URL
  - SUPPORT_CENTER_TEXT
  - TRADEMARK_TEXT
  - FOOTER_LOGO_ALT_TEXT
  - SHOW_FOOTER_LOGO
  - TERMS_OF_SERVICE_URL
  - PRIVACY_POLICY_URL
- Verify the footer links in the MFE
- Verify the Accessibility URL in the Studio MFE

## Screenshots
1. MFE Footer after our changes
<img width="1742" alt="Screenshot 2024-02-22 at 4 32 31 PM" src="https://github.com/openedx/frontend-component-footer/assets/52656433/fc9662e2-32ac-4445-aa62-a3b3077b9c5d">

2. MFE Footer before these changes
<img width="1738" alt="300859871-205c0ce3-8c0a-4c74-b6e9-dde2233af4b3" src="https://github.com/openedx/frontend-component-footer/assets/52656433/dccdb08c-20da-49dc-87f7-39a87df5c650">

3. Legacy LMS Footer:
<img width="1738" alt="300857051-ee5005ec-4a6c-40c6-ba10-18ca61005b79" src="https://github.com/openedx/frontend-component-footer/assets/52656433/0110c209-ac83-4535-b5f2-1146cf74e78b">

4. Studio MFE footer
<img width="1738" alt="300857498-cf882470-bd2a-4997-9202-ecb7cd885ebb" src="https://github.com/openedx/frontend-component-footer/assets/52656433/0284d612-7e7e-4c7a-a8a1-4fb8037ceaa0">

openedx-webhooks commented:
Thanks for the pull request, @asadali145! Please note that it may take us up to several weeks or months to complete a review and merge your PR.

Feel free to add as much of the following information to the ticket as you can:
- supporting documentation
- [Open edX discussion forum threads](https://discuss.openedx.org/)
- timeline information ("this must be merged by XX date", and why that is)
- partner information ("this is a course on edx.org")
- any other information that can help Product understand the context for the PR

All technical communication about the code itself will be done via the GitHub pull request interface. As a reminder, [our process documentation is here](http://edx-developer-guide.readthedocs.org/en/latest/process/overview.html).


This is currently a draft pull request. When it is ready for our review and all tests are green, click "Ready for Review", or remove "WIP" from the title, as appropriate.

<!-- comment:end_of_wip -->




<!-- comment:external_pr -->
<!-- data: eyJkcmFmdCI6IHRydWV9 -->

mphilbrick211 commented:
Hi @asadali145! When you get a moment, would you mind putting a brief description of this contribution at the top of the PR? Thank you!

asadali145 commented:
@mphilbrick211 Sure, I am just testing my changes with a couple of MFEs. Will add the description once I move it to the review.

asadali145 commented:
@mphilbrick211 This is ready now.

pdpinch commented on .env.development:
```diff
@@ -12,7 +12,6 @@ TERMS_OF_SERVICE_URL=null
 PRIVACY_POLICY_URL=null
 SUPPORT_EMAIL=null
 STUDIO_BASE_URL=http://localhost:18010
-SHOW_ACCESSIBILITY_PAGE=false
```
Why did you remove this?

asadali145 replied:
I have tried to make it consistent with other links. The accessibility link was redirecting to `<studio_base_url>/accessibility` if SHOW_ACCESSIBILITY_PAGE=true. I have changed it and added a config ACCESSIBILITY_URL. Now, it will display the link if `ACCESSIBILITY_URL` is set and will redirect to the same URL.

arslanashraf7 replied:
Should we also look for other usages of this variable in [all other MFEs](https://github.com/search?q=org%3Aopenedx%20SHOW_ACCESSIBILITY_PAGE&type=code) that are using it?

arbrandes replied:
Correct.  But this means that this change is a breaking one, so after squashing the commits, the one that remains should be a "feat!:".

pdpinch commented on .env.development:
```diff
@@ -22,3 +21,12 @@ LOGO_URL=https://edx-cdn.org/v3/default/logo.svg
 LOGO_TRADEMARK_URL=https://edx-cdn.org/v3/default/logo-trademark.svg
 LOGO_WHITE_URL=https://edx-cdn.org/v3/default/logo-white.svg
 FAVICON_URL=https://edx-cdn.org/v3/default/favicon.ico
+ACCESSIBILITY_URL=
+ABOUT_US_URL=
+HONOR_CODE_URL=
+CONTACT_URL=
+SUPPORT_CENTER_URL=
+SUPPORT_CENTER_TEXT=
+TRADEMARK_TEXT=
+LOGO_ALT_TEXT=
+SHOW_LOGO=
```
Is there any order to these settings?

pdpinch reviewed (commented):
I'm not too familiar with how this works, but I have a coupel of questions. 

Also, are these settings -- both existing and newly added -- documented anywhere?

asadali145 commented:
> Also, are these settings -- both existing and newly added -- documented anywhere?

@pdpinch A couple of required configs are mentioned in the [readme](https://github.com/openedx/frontend-component-footer?tab=readme-ov-file#environment-variables) but not all. I can add the newly added under a new heading if you want.

arslanashraf7 commented on src/components/Footer.jsx:
```diff
@@ -57,9 +62,29 @@ class SiteFooter extends React.Component {
+            )}
+            <div>
+              <ul className="footer-sub-nav">
+                {this.renderLinkIfExists(process.env.ABOUT_US_URL, 'About Us')}
+                {this.renderLinkIfExists(process.env.TERMS_OF_SERVICE_URL, 'Terms of Service')}
+                {this.renderLinkIfExists(process.env.PRIVACY_POLICY_URL, 'Privacy Policy')}
+                {this.renderLinkIfExists(process.env.HONOR_CODE_URL, 'Honor Code')}
+                {this.renderLinkIfExists(process.env.CONTACT_URL, 'Contact Us')}
+                {this.renderLinkIfExists(process.env.ACCESSIBILITY_URL, 'Accessibility')}
+                {this.renderLinkIfExists(process.env.SUPPORT_CENTER_URL, process.env.SUPPORT_CENTER_TEXT || 'FAQ & Help')}
```
We should use formatMessage with internationalization. An example of that can be seen in [dashboard MFE](https://github.com/openedx/frontend-app-learner-dashboard/blob/master/src/containers/LearnerDashboardHeader/BrandLogo.jsx#L11C11-L11C24)

arslanashraf7 commented on src/components/__snapshots__/Footer.test.jsx.snap:
```diff
@@ -14,7 +14,7 @@ exports[`<Footer /> renders correctly renders with a language selector 1`] = `
       href="http://localhost:18000"
     >
       <img
-        alt="Powered by Open edX"
```
Not sure why we removed this.

arslanashraf7 commented on .env.development:
```diff
@@ -1,8 +1,12 @@
+ABOUT_US_URL=
 ACCESS_TOKEN_COOKIE_NAME=edx-jwt-cookie-header-payload
+ACCESSIBILITY_URL=
```
I think the accessibility is a mandatory page for a website to make it compliant with legalities. I think we should have a default fallback value for this. e.g. how we are making sure in StudioFooter. By default, we are sending the user to STUDIO_URL/accessibility.

arslanashraf7 commented on src/components/studio-footer/StudioFooter.jsx:
```diff
@@ -105,11 +105,8 @@ const StudioFooter = ({
-              data-testid="accessibilityRequest"
-            >
+          {!_.isEmpty(config.ACCESSIBILITY_URL) && (
+            <Hyperlink destination={config.ACCESSIBILITY_URL} data-testid="accessibilityRequest">
```
I left a comment above for a fallback URL for this.

sarina commented:
Could you please upload images in your PR description that aren't broken? I only see a broken image link. Thanks!

pdpinch commented:
@asadali145 I opened a Product Review issue to go with this, https://github.com/openedx/platform-roadmap/issues/324 -- can you add it to the PR description?

ali-hugo commented:
> Could you please upload images in your PR description that aren't broken? I only see a broken image link. Thanks!

@asadali145 +1 to Sarina's request. Also, it would be helpful if you could upload a screencast of the user flow inside Studio (I'd like to perform a product review on this PR, but I'm not a developer so I don't know how to set up a dev environment to do the review). Thanks!

asadali145 commented:
> @asadali145 I opened a Product Review issue to go with this, https://github.com/openedx/platform-roadmap/issues/324 -- can you add it to the PR description?

@pdpinch I have updated the PR description with the Product Review Issue link.

> Could you please upload images in your PR description that aren't broken? I only see a broken image link. Thanks!

@ali-hugo & @sarina I have updated the screenshots. Please check.

> Also, it would be helpful if you could upload a screencast of the user flow inside Studio (I'd like to perform a product review on this PR, but I'm not a developer so I don't know how to set up a dev environment to do the review). Thanks!

@ali-hugo What do you mean by the user flow inside the studio? Would you like me to add Authoring MFE screenshots?

asadali145 commented:
This is ready for review.

CC: @pdpinch @arslanashraf7

ali-hugo commented:
@asadali145 

> I have updated the screenshots. Please check.

Thank you.

> What do you mean by the user flow inside the studio? Would you like me to add Authoring MFE screenshots?

I think that's what I need (sorry, I don't know where the user adds and manages the footer links). Basically, what I would like to see is the process the operator has to follow to add/edit/remove the links (unless this happens inside the code, in which case I won't need to review it). I hope that makes sense!

asadali145 commented:
> I think that's what I need (sorry, I don't know where the user adds and manages the footer links). Basically, what I would like to see is the process the operator has to follow to add/edit/remove the links (unless this happens inside the code, in which case I won't need to review it). I hope that makes sense!

Yeah, now I got your point. In this case, the links are configured through code. DevOps teams can configure these links by adding the required configurations through the code.

ali-hugo commented:
@asadali145 Thanks for confirming that the links are added in the code. 

I am meeting with other members of the product team tomorrow. I would like to get their feedback on this feature before I do my review. I'll get back to you once I have consolidated their feedback.

sarina commented:
@ali-hugo fyi! https://docs.openedx.org/en/latest/developers/references/developer_guide/process/pull-request-statuses.html

jmakowski1123 commented:
Thanks to @pdpinch for pulling this PR into the product review process. I've started the review process here: https://github.com/openedx/platform-roadmap/issues/324#issuecomment-1967850142 - would love for other stakeholders to chime in.

itsjeyd commented:
@ali-hugo It sounds like your original question has been answered (?).

Either way, I thought I'd add that @mphilbrick211 and I currently don't use the `ready for product review` label in the OSPR management workflow. We just add `product review` as a [permanent label](https://github.com/openedx/edx-platform/pull/33952#issuecomment-1956224124) to PRs that need it, and set their status to Product Review on the Contributions board once product review is actually under way 🙂

ali-hugo commented:
@itsjeyd Yes - my original question was answered, thanks.

Thank you for that information about the OSPR workflow. I just want to check something with you about how to find PR's that are ready for a product review. Is the right approach to look for tickets with the "product review" tag in the "Product Review" column [here](https://github.com/orgs/openedx/projects/19/views/1?sliceBy%5Bvalue%5D=Product+Review), and/or in the "Feature Tickets - Product Pull Requests" column [here](https://github.com/orgs/openedx/projects/4/views/20)? Feel free to link me to any documentation if this has already been covered somewhere.

itsjeyd commented:
@ali-hugo I don't think there's a specific process for finding PRs ready for product review (yet). I thought it was in the process of being defined via [this wiki page](https://openedx.atlassian.net/wiki/spaces/COMM/pages/3875962884/How+to+submit+an+open+source+contribution+for+Product+Review#Proposed-process)?

Although that page only talks about what to do for product proposals that the community submits *before* writing any code. It doesn't specify what to do for PRs that need product review but haven't had a proposal submitted first.

You would be able to find those types of PRs via the Product Review column on the Contributions board. And if you're looking for something to review that is smaller in scope, it might make sense to start with them.

If a PR doesn't have a feature ticket on the platform roadmap yet, the first step would be to create one that lists the PR itself as well as any related PRs. The product review process then happens via the feature ticket.

--

At the moment I don't think I can provide any info that's more specific than that. It would be worth further discussing and refining processes with the Core Product Working Group. @mphilbrick211 and I would be happy to help support the workflows that you land on in the context of OSPR management. But we're not necessarily in the driver's seat when it comes to defining them 🙂

arbrandes commented on src/components/Footer.jsx:
```diff
@@ -57,9 +85,43 @@ class SiteFooter extends React.Component {
+          <div className="copyright-col">
+            {getConfig().TRADEMARK_TEXT
+            && (
+            <div className="text-gray-500 small">
```
Shouldn't we be using more semantic class names?  Specially since then it's easier for theming, later.

arbrandes commented on src/components/Footer.messages.js:
```diff
@@ -73,9 +78,14 @@ const messages = defineMessages({
+  'footer.legalLinks.honorCode': {
+    id: 'footer.legalLinks.honorCode',
+    defaultMessage: 'Honor Code',
+    description: 'The label for the link to the edX honor code page.',
```
```suggestion
    description: 'The label for the link to the honor code page.',
```

And if you don't mind applying this to the other messages, it would be great. :)

asadali145 replied:
Just to confirm, Should I apply this change to all the existing messages?

arbrandes replied:
It would be awesome if you did, yes.  Not a requirement, though.

asadali145 replied:
I have changed a few. Let me know if you want me to change other messages like the social media app messages.

arbrandes reviewed (changes requested):
Looks great, thanks!  But can you please address comments, including squashing all commits into one with a leading `feat!` or even better, a `BREAKING CHANGE` comment explaining what users of the new version need to be on the lookout for?

asadali145 commented:
@arbrandes Could you please have another look?

arbrandes commented on src/components/Footer.jsx:
```diff
@@ -57,9 +85,43 @@ class SiteFooter extends React.Component {
+          <div className="copyright-col">
+            {getConfig().TRADEMARK_TEXT
+            && (
+            <div className="trademark-text text-gray-500 small">
```
I still think `text-gray-500` is too specific to be hard-coded.  Maybe remove it and reimplement it in the CSS, so it can be overriden more explicitly by theming later?

asadali145 replied:
This was a bootstrap mixin. I have updated it now.

arbrandes commented:
Ok, I'm going to actually try it out shortly, and if it works as expected I'll approve!

ali-hugo commented:
@itsjeyd Thank you for the detailed explanation. 

> Although that page only talks about what to do for product proposals that the community submits _before_ writing any code. It doesn't specify what to do for PRs that need product review but haven't had a proposal submitted first.

Yes, that's where my confusion comes in. I will broach this topic with the Core Product Working Group and see if we can get a procedure put in place for reviewing PR's that don't have product proposals. At some point I may pull you and Michelle into the conversation for your recommendations.

Thanks for your help on this.

pdpinch commented:
@itsjeyd @ali-hugo I opened https://github.com/openedx/platform-roadmap/issues/324 -- does that qualify as a product proposal?

sarina commented:
> "needs-product-review-but-doesn't-have-a-proposal" category

I'll double-check but I think it's very simple: if it needs product review, everyone should stop (code reviewing/developing) the PR, make the proposal, and go through the product review process. It's a waste of (code/pr) reviewer's time to be looking at something that doesn't have product review.

itsjeyd commented:
@pdpinch 

> I opened https://github.com/openedx/platform-roadmap/issues/324 -- does that qualify as a product proposal?

The feature ticket LGTM, thanks for creating it. I'm not part of the product working group, though, and don't have all the necessary context to answer your question definitively.

CC @ali-hugo @mphilbrick211

ali-hugo commented:
@arbrandes 

> @ali-hugo, please let me know what the proper procedure ends up being: I expect to be finding a bunch of PRs in the "needs-product-review-but-doesn't-have-a-proposal" category in the near future.

We're in the process of defining the procedure [in this document](https://openedx.atlassian.net/wiki/spaces/COMM/pages/3875962884/How+to+submit+an+open+source+contribution+for+Product+Review?focusedCommentId=4093575271). Please keep an eye on that to see how the decisions unfold. Feel free to add any comments there too.

asadali145 commented:
> Ok, I'm going to actually try it out shortly, and if it works as expected I'll approve!

@arbrandes Do you have an update on testing? I have rebased my PR to resolve a few conflicts.

arbrandes reviewed (changes requested):
I think that at least for now we're going to go on a different architectural direction.  As I've been mentioning elsewhere, the footer (and of course, the header) are perfect candidates for plugin slots: this will allow operators to customize them much more flexibly than just setting a few hard-coded links.

This effort is already in progress via https://github.com/openedx/wg-frontend/issues/178, and we're hoping to make it available in time for the Redwood release.

For the record, there is a competing implementation of the footer as a slot that _also_ allows individual links to be configured via a backend API call.  I figure if we're going with this level of granularity at all, that PR is a better approach:

https://github.com/openedx/frontend-component-footer/pull/405

arbrandes commented:
As promised, here's the thread:

https://discuss.openedx.org/t/plugin-slots-vs-configuration/13009

itsjeyd commented:
@asadali145 @arbrandes It sounds like this PR is no longer needed. Unless you have any objections, I'll close it next week.

asadali145 commented:
@arbrandes Thanks for all of the work around the Plugin Slots.

We were waiting for the Redwood release to test the FooterSlot. Here are a few things that we have noticed:

- The FooterSlot(https://github.com/openedx/frontend-slot-footer) uses(https://github.com/openedx/frontend-slot-footer/blob/master/src/FooterSlot.jsx) the same old footer component from https://github.com/openedx/frontend-component-footer.
- The learning footer does not support any links in the Footer. The [StudioFooter](https://github.com/openedx/frontend-component-footer/blob/master/src/components/studio-footer/StudioFooter.jsx) does support a few links i.e. Privacy, Terms of Service.
- I went through your comments on My PR, https://github.com/openedx/frontend-component-footer/pull/405, https://github.com/openedx/frontend-component-footer/pull/326, and https://discuss.openedx.org/t/plugin-slots-vs-configuration/13009. I see you have a few reservations about using the Config API, Adding a new API for Footer Configs, and adding environment variables like in this PR.
- As the Learning Footer component still has no support for the links, I have a question:
  - Do we want to support custom links in the Footer Component? If so, please suggest an approach.

CC: @pdpinch

arbrandes commented:
So, as per recent conversations with @brian-smith-tcril and @davidjoy, going forward we're going to be ok with a configuration-controlled header and footer (beyond just plugin slots), _provided the configuration mechanism is vetted and consistently applied_, which is definitely not the case with the `.env` variables that exist now.

To summarize a few key consequences of the above:

* This sane configuration mechanism doesn't yet exist, but is being worked on by @davidjoy as part of the OEP-65/Module Federation work that's currently targetting Teak.
* We won't accept drive-by PRs adding configuration: each new configuration structure will have to undergo the [product proposal and review process](https://openedx.atlassian.net/wiki/spaces/COMM/pages/3875962884/How+to+submit+an+open+source+contribution+for+Product+Review), with a proper proposal and justification such as [the Header's](https://openedx.atlassian.net/wiki/spaces/OEPM/pages/4412243977/Proposal+Customizable+Header+for+LMS).

It would be a great benefit to the community if you wrote just such a proposal for the footer.  The sooner it gets reviewed and approved, the higher the chance that we can convince David to include some of it as he works on frontend-base (where both Header and Footer will live, going forward).

asadali145 commented:
@arbrandes Thanks for the detailed reply. I have written a proposal for the footer https://openedx.atlassian.net/wiki/spaces/OEPM/pages/4512514079/Proposal+Customizable+Footer+for+LMS+CMS.

sarina commented:
Hi everyone! The product proposal has been accepted. What's the next steps for this PR?

asadali145 commented:
> Hi everyone! The product proposal has been accepted. What's the next steps for this PR?

This can be closed. I am closing it.
